Bug with Python 3.12

As of now, several components break with Python 3.12. Here is a sample of the output after installing Python 3.12 in the datatree dev environment. It appears to be related to pathlib, but I am not exactly sure. Any thoughts here?

```python-traceback
self = <[AttributeError("'NodePath' object has no attribute '_raw_paths'") raised in repr()] NodePath object at 0x1347ae3d0>

    @property
    def drive(self):
        """The drive prefix (letter or UNC path), if any."""
        try:
>           return self._drv
E           AttributeError: 'NodePath' object has no attribute '_drv'

../../mambaforge/envs/datatree-test/lib/python3.12/pathlib.py:555: AttributeError

During handling of the above exception, another exception occurred:

self = <datatree.tests.test_treenode.TestGetNodes object at 0x13480f020>

    def test_get_upwards(self):
        sue = TreeNode()
        kate = TreeNode()
        mary = TreeNode(children={"Sue": sue, "Kate": kate})
        john = TreeNode(children={"Mary": mary})

>       assert sue._get_item("../") is mary

datatree/tests/test_treenode.py:132:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
datatree/treenode.py:429: in _get_item
    path = NodePath(path)
datatree/treenode.py:36: in __new__
    if obj.drive:
../../mambaforge/envs/datatree-test/lib/python3.12/pathlib.py:557: in drive
    self._load_parts()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <[AttributeError("'NodePath' object has no attribute '_raw_paths'") raised in repr()] NodePath object at 0x1347ae3d0>

    def _load_parts(self):
>       paths = self._raw_paths
E       AttributeError: 'NodePath' object has no attribute '_raw_paths'

../../mambaforge/envs/datatree-test/lib/python3.12/pathlib.py:408: AttributeError
_______________________ TestGetNodes.test_get_from_root ________________________

self = <[AttributeError("'NodePath' object has no attribute '_raw_paths'") raised in repr()] NodePath object at 0x1359a9a50>

    @property
    def drive(self):
        """The drive prefix (letter or UNC path), if any."""
        try:
>           return self._drv
E           AttributeError: 'NodePath' object has no attribute '_drv'

../../mambaforge/envs/datatree-test/lib/python3.12/pathlib.py:555: AttributeError

During handling of the above exception, another exception occurred:

self = <datatree.tests.test_treenode.TestGetNodes object at 0x13480f320>

    def test_get_from_root(self):
        sue = TreeNode()
        mary = TreeNode(children={"Sue": sue})
        john = TreeNode(children={"Mary": mary})  # noqa

>       assert sue._get_item("/Mary") is mary

datatree/tests/test_treenode.py:143:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
datatree/treenode.py:429: in _get_item
    path = NodePath(path)
datatree/treenode.py:36: in __new__
    if obj.drive:
../../mambaforge/envs/datatree-test/lib/python3.12/pathlib.py:557: in drive
    self._load_parts()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <[AttributeError("'NodePath' object has no attribute '_raw_paths'") raised in repr()] NodePath object at 0x1359a9a50>

    def _load_parts(self):
>       paths = self._raw_paths
E       AttributeError: 'NodePath' object has no attribute '_raw_paths'

../../mambaforge/envs/datatree-test/lib/python3.12/pathlib.py:408: AttributeError
_______________________ TestSetNodes.test_set_child_node _______________________

self = <[AttributeError("'NodePath' object has no attribute '_raw_paths'") raised in repr()] NodePath object at 0x135a14f50>

    @property
    def drive(self):
        """The drive prefix (letter or UNC path), if any."""
        try:
>           return self._drv
E           AttributeError: 'NodePath' object has no attribute '_drv'

../../mambaforge/envs/datatree-test/lib/python3.12/pathlib.py:555: AttributeError

During handling of the above exception, another exception occurred:

self = <datatree.tests.test_treenode.TestSetNodes object at 0x13480e780>

    def test_set_child_node(self):
        john = TreeNode()
        mary = TreeNode()
>       john._set_item("Mary", mary)

datatree/tests/test_treenode.py:150:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
datatree/treenode.py:491: in _set_item
    path = NodePath(path)
datatree/treenode.py:36: in __new__
    if obj.drive:
../../mambaforge/envs/datatree-test/lib/python3.12/pathlib.py:557: in drive
    self._load_parts()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <[AttributeError("'NodePath' object has no attribute '_raw_paths'") raised in repr()] NodePath object at 0x135a14f50>

    def _load_parts(self):
>       paths = self._raw_paths
E       AttributeError: 'NodePath' object has no attribute '_raw_paths'

../../mambaforge/envs/datatree-test/lib/python3.12/pathlib.py:408: AttributeError
____________________ TestSetNodes.test_child_already_exists ____________________

self = <[AttributeError("'NodePath' object has no attribute '_raw_paths'") raised in repr()] NodePath object at 0x1359fca50>

    @property
    def drive(self):
        """The drive prefix (letter or UNC path), if any."""
        try:
>           return self._drv
E           AttributeError: 'NodePath' object has no attribute '_drv'

../../mambaforge/envs/datatree-test/lib/python3.12/pathlib.py:555: AttributeError

During handling of the above exception, another exception occurred:

self = <datatree.tests.test_treenode.TestSetNodes object at 0x13480e1b0>

    def test_child_already_exists(self):
        mary = TreeNode()
        john = TreeNode(children={"Mary": mary})
        mary_2 = TreeNode()
        with pytest.raises(KeyError):
>           john._set_item("Mary", mary_2, allow_overwrite=False)

datatree/tests/test_treenode.py:162:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
datatree/treenode.py:491: in _set_item
    path = NodePath(path)
datatree/treenode.py:36: in __new__
    if obj.drive:
../../mambaforge/envs/datatree-test/lib/python3.12/pathlib.py:557: in drive
    self._load_parts()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <[AttributeError("'NodePath' object has no attribute '_raw_paths'") raised in repr()] NodePath object at 0x1359fca50>

    def _load_parts(self):
>       paths = self._raw_paths
E       AttributeError: 'NodePath' object has no attribute '_raw_paths'

../../mambaforge/envs/datatree-test/lib/python3.12/pathlib.py:408: AttributeError
_______________________ TestSetNodes.test_set_grandchild _______________________

self = <[AttributeError("'NodePath' object has no attribute '_raw_paths'") raised in repr()] NodePath object at 0x1346ea450>

    @property
    def drive(self):
        """The drive prefix (letter or UNC path), if any."""
        try:
>           return self._drv
E           AttributeError: 'NodePath' object has no attribute '_drv'

../../mambaforge/envs/datatree-test/lib/python3.12/pathlib.py:555: AttributeError

During handling of the above exception, another exception occurred:

self = <datatree.tests.test_treenode.TestSetNodes object at 0x13480e300>

    def test_set_grandchild(self):
        rose = TreeNode()
        mary = TreeNode()
        john = TreeNode()

>       john._set_item("Mary", mary)

datatree/tests/test_treenode.py:169:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
datatree/treenode.py:491: in _set_item
    path = NodePath(path)
datatree/treenode.py:36: in __new__
    if obj.drive:
../../mambaforge/envs/datatree-test/lib/python3.12/pathlib.py:557: in drive
    self._load_parts()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <[AttributeError("'NodePath' object has no attribute '_raw_paths'") raised in repr()] NodePath object at 0x1346ea450>

    def _load_parts(self):
>       paths = self._raw_paths
E       AttributeError: 'NodePath' object has no attribute '_raw_paths'

../../mambaforge/envs/datatree-test/lib/python3.12/pathlib.py:408: AttributeError
_________________ TestSetNodes.test_create_intermediate_child __________________

self = <[AttributeError("'NodePath' object has no attribute '_raw_paths'") raised in repr()] NodePath object at 0x1347ae3d0>

    @property
    def drive(self):
        """The drive prefix (letter or UNC path), if any."""
        try:
>           return self._drv
E           AttributeError: 'NodePath' object has no attribute '_drv'

../../mambaforge/envs/datatree-test/lib/python3.12/pathlib.py:555: AttributeError

During handling of the above exception, another exception occurred:

self = <datatree.tests.test_treenode.TestSetNodes object at 0x13480d8e0>

    def test_create_intermediate_child(self):
        john = TreeNode()
        rose = TreeNode()

        # test intermediate children not allowed
        with pytest.raises(KeyError, match="Could not reach"):
>           john._set_item(path="Mary/Rose", item=rose, new_nodes_along_path=False)

datatree/tests/test_treenode.py:183:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
datatree/treenode.py:491: in _set_item
    path = NodePath(path)
datatree/treenode.py:36: in __new__
    if obj.drive:
../../mambaforge/envs/datatree-test/lib/python3.12/pathlib.py:557: in drive
    self._load_parts()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <[AttributeError("'NodePath' object has no attribute '_raw_paths'") raised in repr()] NodePath object at 0x1347ae3d0>

    def _load_parts(self):
>       paths = self._raw_paths
E       AttributeError: 'NodePath' object has no attribute '_raw_paths'

../../mambaforge/envs/datatree-test/lib/python3.12/pathlib.py:408: AttributeError
______________________ TestSetNodes.test_overwrite_child _______________________

self = <[AttributeError("'NodePath' object has no attribute '_raw_paths'") raised in repr()] NodePath object at 0x134bc3650>

    @property
    def drive(self):
        """The drive prefix (letter or UNC path), if any."""
        try:
>           return self._drv
E           AttributeError: 'NodePath' object has no attribute '_drv'

../../mambaforge/envs/datatree-test/lib/python3.12/pathlib.py:555: AttributeError

During handling of the above exception, another exception occurred:

self = <datatree.tests.test_treenode.TestSetNodes object at 0x13480d910>

    def test_overwrite_child(self):
        john = TreeNode()
        mary = TreeNode()
>       john._set_item("Mary", mary)

datatree/tests/test_treenode.py:197:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
datatree/treenode.py:491: in _set_item
    path = NodePath(path)
datatree/treenode.py:36: in __new__
    if obj.drive:
../../mambaforge/envs/datatree-test/lib/python3.12/pathlib.py:557: in drive
    self._load_parts()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <[AttributeError("'NodePath' object has no attribute '_raw_paths'") raised in repr()] NodePath object at 0x134bc3650>

    def _load_parts(self):
>       paths = self._raw_paths
E       AttributeError: 'NodePath' object has no attribute '_raw_paths'

../../mambaforge/envs/datatree-test/lib/python3.12/pathlib.py:408: AttributeError
__________________________ TestPruning.test_del_child __________________________

self = <[AttributeError("'NodePath' object has no attribute '_raw_paths'") raised in repr()] NodePath object at 0x134215750>

    @property
    def drive(self):
        """The drive prefix (letter or UNC path), if any."""
        try:
>           return self._drv
E           AttributeError: 'NodePath' object has no attribute '_drv'

../../mambaforge/envs/datatree-test/lib/python3.12/pathlib.py:555: AttributeError

During handling of the above exception, another exception occurred:

self = <datatree.tests.test_treenode.TestPruning object at 0x13480dc40>

    def test_del_child(self):
        john = TreeNode()
        mary = TreeNode()
>       john._set_item("Mary", mary)

datatree/tests/test_treenode.py:217:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
datatree/treenode.py:491: in _set_item
    path = NodePath(path)
datatree/treenode.py:36: in __new__
    if obj.drive:
../../mambaforge/envs/datatree-test/lib/python3.12/pathlib.py:557: in drive
    self._load_parts()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <[AttributeError("'NodePath' object has no attribute '_raw_paths'") raised in repr()] NodePath object at 0x134215750>

    def _load_parts(self):
>       paths = self._raw_paths
E       AttributeError: 'NodePath' object has no attribute '_raw_paths'

../../mambaforge/envs/datatree-test/lib/python3.12/pathlib.py:408: AttributeError
=============================== warnings summary ===============================
../../mambaforge/envs/datatree-test/lib/python3.12/site-packages/dateutil/tz/tz.py:37
  /Users/mgrover/mambaforge/envs/datatree-test/lib/python3.12/site-packages/dateutil/tz/tz.py:37: DeprecationWarning: datetime.datetime.utcfromtimestamp() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.fromtimestamp(timestamp, datetime.UTC).
    EPOCH = datetime.datetime.utcfromtimestamp(0)

../../mambaforge/envs/datatree-test/lib/python3.12/site-packages/xarray/tests/__init__.py:133
  /Users/mgrover/mambaforge/envs/datatree-test/lib/python3.12/site-packages/xarray/tests/__init__.py:133: PytestUnknownMarkWarning: Unknown pytest.mark.flaky - is this a typo?  You can register custom marks to avoid this warning - for details, see https://docs.pytest.org/en/stable/how-to/mark.html
    flaky = pytest.mark.flaky

../../mambaforge/envs/datatree-test/lib/python3.12/site-packages/xarray/tests/__init__.py:134
  /Users/mgrover/mambaforge/envs/datatree-test/lib/python3.12/site-packages/xarray/tests/__init__.py:134: PytestUnknownMarkWarning: Unknown pytest.mark.network - is this a typo?  You can register custom marks to avoid this warning - for details, see https://docs.pytest.org/en/stable/how-to/mark.html
    network = pytest.mark.network

datatree/tests/test_io.py::TestIO::test_to_zarr
  <frozen posixpath>:82: RuntimeWarning: deallocating CachingFileManager(<class 'netCDF4._netCDF4.Dataset'>, '/private/var/folders/bw/c9j8z20x45s2y20vv6528qjc0000gq/T/pytest-of-mgrover/pytest-0/test_to_h5netcdf0/test.nc', mode='r', kwargs={'clobber': True, 'diskless': False, 'persist': False, 'format': 'NETCDF4'}, manager_id='1db2a12b-6ca7-4f13-84cd-21bb6d3e8f57'), but file is not already closed. This may indicate a bug.

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
=========================== short test summary info ============================
FAILED datatree/tests/test_dataset_api.py::TestDSMethodInheritance::test_dataset_method - AttributeError: 'NodePath' object has no attribute '_raw_paths'
FAILED datatree/tests/test_dataset_api.py::TestDSMethodInheritance::test_reduce_method - AttributeError: 'NodePath' object has no attribute '_raw_paths'
FAILED datatree/tests/test_dataset_api.py::TestDSMethodInheritance::test_nan_reduce_method - AttributeError: 'NodePath' object has no attribute '_raw_paths'
FAILED datatree/tests/test_dataset_api.py::TestDSMethodInheritance::test_cum_method - AttributeError: 'NodePath' object has no attribute '_raw_paths'
FAILED datatree/tests/test_dataset_api.py::TestOps::test_binary_op_on_int - AttributeError: 'NodePath' object has no attribute '_raw_paths'
FAILED datatree/tests/test_dataset_api.py::TestOps::test_binary_op_on_dataset - AttributeError: 'NodePath' object has no attribute '_raw_paths'
FAILED datatree/tests/test_dataset_api.py::TestOps::test_binary_op_on_datatree - AttributeError: 'NodePath' object has no attribute '_raw_paths'
FAILED datatree/tests/test_dataset_api.py::TestUFuncs::test_tree - AttributeError: 'NodePath' object has no attribute '_raw_paths'
FAILED datatree/tests/test_datatree.py::TestPaths::test_path_roundtrip - AttributeError: 'NodePath' object has no attribute '_raw_paths'
FAILED datatree/tests/test_datatree.py::TestPaths::test_relative_paths - AttributeError: 'NodePath' object has no attribute '_raw_paths'
FAILED datatree/tests/test_datatree.py::TestGetItem::test_getitem_node - AttributeError: 'NodePath' object has no attribute '_raw_paths'
FAILED datatree/tests/test_datatree.py::TestGetItem::test_getitem_self - AttributeError: 'NodePath' object has no attribute '_raw_paths'
FAILED datatree/tests/test_datatree.py::TestGetItem::test_getitem_single_data_variable - AttributeError: 'NodePath' object has no attribute '_raw_paths'
FAILED datatree/tests/test_datatree.py::TestGetItem::test_getitem_single_data_variable_from_node - AttributeError: 'NodePath' object has no attribute '_raw_paths'
FAILED datatree/tests/test_datatree.py::TestGetItem::test_getitem_nonexistent_node - AttributeError: 'NodePath' object has no attribute '_raw_paths'
FAILED datatree/tests/test_datatree.py::TestGetItem::test_getitem_nonexistent_variable - AttributeError: 'NodePath' object has no attribute '_raw_paths'
FAILED datatree/tests/test_datatree.py::TestUpdate::test_update - AttributeError: 'NodePath' object has no attribute '_raw_paths'
FAILED datatree/tests/test_datatree.py::TestUpdate::test_update_new_named_dataarray - AttributeError: 'NodePath' object has no attribute '_raw_paths'
FAILED datatree/tests/test_datatree.py::TestUpdate::test_update_doesnt_alter_child_name - AttributeError: 'NodePath' object has no attribute '_raw_paths'
FAILED datatree/tests/test_datatree.py::TestUpdate::test_update_overwrite - AttributeError: 'NodePath' object has no attribute '_raw_paths'
FAILED datatree/tests/test_datatree.py::TestCopy::test_copy - AttributeError: 'NodePath' object has no attribute '_raw_paths'
FAILED datatree/tests/test_datatree.py::TestCopy::test_copy_subtree - AttributeError: 'NodePath' object has no attribute '_raw_paths'
FAILED datatree/tests/test_datatree.py::TestCopy::test_deepcopy - AttributeError: 'NodePath' object has no attribute '_raw_paths'
FAILED datatree/tests/test_datatree.py::TestSetItem::test_setitem_new_child_node - AttributeError: 'NodePath' object has no attribute '_raw_paths'
FAILED datatree/tests/test_datatree.py::TestSetItem::test_setitem_unnamed_child_node_becomes_named - AttributeError: 'NodePath' object has no attribute '_raw_paths'
FAILED datatree/tests/test_datatree.py::TestSetItem::test_setitem_new_grandchild_node - AttributeError: 'NodePath' object has no attribute '_raw_paths'
FAILED datatree/tests/test_datatree.py::TestSetItem::test_grafted_subtree_retains_name - AttributeError: 'NodePath' object has no attribute '_raw_paths'
FAILED datatree/tests/test_datatree.py::TestSetItem::test_setitem_new_empty_node - AttributeError: 'NodePath' object has no attribute '_raw_paths'
FAILED datatree/tests/test_datatree.py::TestSetItem::test_setitem_overwrite_data_in_node_with_none - AttributeError: 'NodePath' object has no attribute '_raw_paths'
FAILED datatree/tests/test_datatree.py::TestSetItem::test_setitem_named_dataarray - AttributeError: 'NodePath' object has no attribute '_raw_paths'
FAILED datatree/tests/test_datatree.py::TestSetItem::test_setitem_unnamed_dataarray - AttributeError: 'NodePath' object has no attribute '_raw_paths'
FAILED datatree/tests/test_datatree.py::TestSetItem::test_setitem_variable - AttributeError: 'NodePath' object has no attribute '_raw_paths'
FAILED datatree/tests/test_datatree.py::TestSetItem::test_setitem_coerce_to_dataarray - AttributeError: 'NodePath' object has no attribute '_raw_paths'
FAILED datatree/tests/test_datatree.py::TestSetItem::test_setitem_add_new_variable_to_empty_node - AttributeError: 'NodePath' object has no attribute '_raw_paths'
FAILED datatree/tests/test_datatree.py::TestSetItem::test_setitem_dataarray_replace_existing_node - AttributeError: 'NodePath' object has no attribute '_raw_paths'
FAILED datatree/tests/test_datatree.py::TestTreeFromDict::test_one_layer - AttributeError: 'NodePath' object has no attribute '_raw_paths'
FAILED datatree/tests/test_datatree.py::TestTreeFromDict::test_two_layers - AttributeError: 'NodePath' object has no attribute '_raw_paths'
FAILED datatree/tests/test_datatree.py::TestTreeFromDict::test_nones - AttributeError: 'NodePath' object has no attribute '_raw_paths'
FAILED datatree/tests/test_datatree.py::TestTreeFromDict::test_datatree_values - AttributeError: 'NodePath' object has no attribute '_raw_paths'
FAILED datatree/tests/test_datatree.py::TestTreeFromDict::test_roundtrip - AttributeError: 'NodePath' object has no attribute '_raw_paths'
FAILED datatree/tests/test_datatree.py::TestDatasetView::test_arithmetic - AttributeError: 'NodePath' object has no attribute '_raw_paths'
FAILED datatree/tests/test_datatree.py::TestAccess::test_attribute_access - AttributeError: 'NodePath' object has no attribute '_raw_paths'
FAILED datatree/tests/test_datatree.py::TestRestructuring::test_drop_nodes - AttributeError: 'NodePath' object has no attribute '_raw_paths'
FAILED datatree/tests/test_datatree.py::TestRestructuring::test_assign - AttributeError: 'NodePath' object has no attribute '_raw_paths'
FAILED datatree/tests/test_datatree.py::TestPipe::test_params - AttributeError: 'NodePath' object has no attribute '_raw_paths'
FAILED datatree/tests/test_datatree.py::TestSubset::test_filter - AttributeError: 'NodePath' object has no attribute '_raw_paths'
FAILED datatree/tests/test_formatting.py::TestDiffFormatting::test_diff_structure - AttributeError: 'NodePath' object has no attribute '_raw_paths'
FAILED datatree/tests/test_formatting.py::TestDiffFormatting::test_diff_node_names - AttributeError: 'NodePath' object has no attribute '_raw_paths'
FAILED datatree/tests/test_formatting.py::TestDiffFormatting::test_diff_node_data - AttributeError: 'NodePath' object has no attribute '_raw_paths'
FAILED datatree/tests/test_io.py::TestIO::test_to_netcdf - AttributeError: 'NodePath' object has no attribute '_raw_paths'
FAILED datatree/tests/test_io.py::TestIO::test_netcdf_encoding - AttributeError: 'NodePath' object has no attribute '_raw_paths'
FAILED datatree/tests/test_io.py::TestIO::test_to_h5netcdf - AttributeError: 'NodePath' object has no attribute '_raw_paths'
FAILED datatree/tests/test_io.py::TestIO::test_to_zarr - AttributeError: 'NodePath' object has no attribute '_raw_paths'
FAILED datatree/tests/test_io.py::TestIO::test_zarr_encoding - AttributeError: 'NodePath' object has no attribute '_raw_paths'
FAILED datatree/tests/test_io.py::TestIO::test_to_zarr_zip_store - AttributeError: 'NodePath' object has no attribute '_raw_paths'
FAILED datatree/tests/test_io.py::TestIO::test_to_zarr_not_consolidated - AttributeError: 'NodePath' object has no attribute '_raw_paths'
FAILED datatree/tests/test_mapping.py::TestCheckTreesIsomorphic::test_different_widths - AttributeError: 'NodePath' object has no attribute '_raw_paths'
FAILED datatree/tests/test_mapping.py::TestCheckTreesIsomorphic::test_different_heights - AttributeError: 'NodePath' object has no attribute '_raw_paths'
FAILED datatree/tests/test_mapping.py::TestCheckTreesIsomorphic::test_names_different - AttributeError: 'NodePath' object has no attribute '_raw_paths'
FAILED datatree/tests/test_mapping.py::TestCheckTreesIsomorphic::test_isomorphic_names_equal - AttributeError: 'NodePath' object has no attribute '_raw_paths'
FAILED datatree/tests/test_mapping.py::TestCheckTreesIsomorphic::test_isomorphic_ordering - AttributeError: 'NodePath' object has no attribute '_raw_paths'
FAILED datatree/tests/test_mapping.py::TestCheckTreesIsomorphic::test_isomorphic_names_not_equal - AttributeError: 'NodePath' object has no attribute '_raw_paths'
FAILED datatree/tests/test_mapping.py::TestCheckTreesIsomorphic::test_not_isomorphic_complex_tree - AttributeError: 'NodePath' object has no attribute '_raw_paths'
FAILED datatree/tests/test_mapping.py::TestMapOverSubTree::test_not_isomorphic - AttributeError: 'NodePath' object has no attribute '_raw_paths'
FAILED datatree/tests/test_mapping.py::TestMapOverSubTree::test_single_dt_arg - AttributeError: 'NodePath' object has no attribute '_raw_paths'
FAILED datatree/tests/test_mapping.py::TestMapOverSubTree::test_single_dt_arg_plus_args_and_kwargs - AttributeError: 'NodePath' object has no attribute '_raw_paths'
FAILED datatree/tests/test_mapping.py::TestMapOverSubTree::test_multiple_dt_args - AttributeError: 'NodePath' object has no attribute '_raw_paths'
FAILED datatree/tests/test_mapping.py::TestMapOverSubTree::test_dt_as_kwarg - AttributeError: 'NodePath' object has no attribute '_raw_paths'
FAILED datatree/tests/test_mapping.py::TestMapOverSubTree::test_return_multiple_dts - AttributeError: 'NodePath' object has no attribute '_raw_paths'
FAILED datatree/tests/test_mapping.py::TestMapOverSubTree::test_map_single_dataset_against_whole_tree - AttributeError: 'NodePath' object has no attribute '_raw_paths'
FAILED datatree/tests/test_mapping.py::TestMapOverSubTree::test_dt_method - AttributeError: 'NodePath' object has no attribute '_raw_paths'
FAILED datatree/tests/test_mapping.py::TestMapOverSubTree::test_discard_ancestry - AttributeError: 'NodePath' object has no attribute '_raw_paths'
FAILED datatree/tests/test_mapping.py::TestMutableOperations::test_construct_using_type - AttributeError: 'NodePath' object has no attribute '_raw_paths'
FAILED datatree/tests/test_mapping.py::TestMutableOperations::test_alter_inplace - AttributeError: 'NodePath' object has no attribute '_raw_paths'
FAILED datatree/tests/test_treenode.py::TestGetNodes::test_get_child - AttributeError: 'NodePath' object has no attribute '_raw_paths'
FAILED datatree/tests/test_treenode.py::TestGetNodes::test_get_upwards - AttributeError: 'NodePath' object has no attribute '_raw_paths'
FAILED datatree/tests/test_treenode.py::TestGetNodes::test_get_from_root - AttributeError: 'NodePath' object has no attribute '_raw_paths'
FAILED datatree/tests/test_treenode.py::TestSetNodes::test_set_child_node - AttributeError: 'NodePath' object has no attribute '_raw_paths'
FAILED datatree/tests/test_treenode.py::TestSetNodes::test_child_already_exists - AttributeError: 'NodePath' object has no attribute '_raw_paths'
FAILED datatree/tests/test_treenode.py::TestSetNodes::test_set_grandchild - AttributeError: 'NodePath' object has no attribute '_raw_paths'
FAILED datatree/tests/test_treenode.py::TestSetNodes::test_create_intermediate_child - AttributeError: 'NodePath' object has no attribute '_raw_paths'
FAILED datatree/tests/test_treenode.py::TestSetNodes::test_overwrite_child - AttributeError: 'NodePath' object has no attribute '_raw_paths'
FAILED datatree/tests/test_treenode.py::TestPruning::test_del_child - AttributeError: 'NodePath' object has no attribute '_raw_paths'
```

--- a/.github/workflows/main.yaml
+++ b/.github/workflows/main.yaml
@@ -48,7 +48,7 @@ jobs:
           python -m pytest --cov=./ --cov-report=xml --verbose
 
       - name: Upload code coverage to Codecov
-        uses: codecov/codecov-action@v3.1.1
+        uses: codecov/codecov-action@v3.1.4
         with:
           file: ./coverage.xml
           flags: unittests

--- a/.github/workflows/main.yaml
+++ b/.github/workflows/main.yaml
@@ -65,7 +65,7 @@ jobs:
         shell: bash -l {0}
     strategy:
       matrix:
-        python-version: ["3.9", "3.10", "3.11"]
+        python-version: ["3.9", "3.10", "3.11", "3.12"]
     steps:
       - uses: actions/checkout@v4
 

--- a/.github/workflows/main.yaml
+++ b/.github/workflows/main.yaml
@@ -20,7 +20,7 @@ jobs:
         shell: bash -l {0}
     strategy:
       matrix:
-        python-version: ["3.9", "3.10", "3.11"]
+        python-version: ["3.9", "3.10", "3.11", "3.12"]
     steps:
       - uses: actions/checkout@v3
 

--- a/datatree/tests/test_treenode.py
+++ b/datatree/tests/test_treenode.py
@@ -1,7 +1,7 @@
 import pytest
 
 from datatree.iterators import LevelOrderIter, PreOrderIter
-from datatree.treenode import InvalidTreeError, NamedNode, TreeNode
+from datatree.treenode import InvalidTreeError, NamedNode, NodePath, TreeNode
 
 
 class TestFamilyTree:
@@ -369,3 +369,9 @@ class TestRenderTree:
         ]
         for expected_node, printed_node in zip(expected_nodes, printout.splitlines()):
             assert expected_node in printed_node
+
+
+def test_nodepath():
+    path = NodePath("/Mary")
+    assert path.root == "/"
+    assert path.stem == "Mary"

--- a/datatree/treenode.py
+++ b/datatree/treenode.py
@@ -29,6 +29,7 @@ class NotFoundInTreeError(ValueError):
 
 
 class NodePath(PurePosixPath):
+    """Represents a path from one node to another within a tree."""
     def __init__(self, *pathsegments):
         if sys.version_info >= (3, 12):
             super().__init__(*pathsegments)
@@ -41,7 +42,7 @@ class NodePath(PurePosixPath):
             raise ValueError(
                 'Root of NodePath can only be either "/" or "", with "" meaning the path is relative.'
             )
-
+        # TODO should we also forbid suffixes to avoid node names with dots in them?
 
 Tree = TypeVar("Tree", bound="TreeNode")
 

--- a/datatree/treenode.py
+++ b/datatree/treenode.py
@@ -1,5 +1,6 @@
 from __future__ import annotations
 
+import sys
 from collections import OrderedDict
 from pathlib import PurePosixPath
 from typing import (
@@ -28,22 +29,18 @@ class NotFoundInTreeError(ValueError):
 
 
 class NodePath(PurePosixPath):
-    """Represents a path from one node to another within a tree."""
-
-    def __new__(cls, *args: str | "NodePath") -> "NodePath":
-        obj = super().__new__(cls, *args)
-
-        if obj.drive:
+    def __init__(self, *pathsegments):
+        if sys.version_info >= (3, 12):
+            super().__init__(*pathsegments)
+        else:
+            super().__new__(PurePosixPath, *pathsegments)
+        if self.drive:
             raise ValueError("NodePaths cannot have drives")
 
-        if obj.root not in ["/", ""]:
+        if self.root not in ["/", ""]:
             raise ValueError(
                 'Root of NodePath can only be either "/" or "", with "" meaning the path is relative.'
             )
-
-        # TODO should we also forbid suffixes to avoid node names with dots in them?
-
-        return obj
 
 
 Tree = TypeVar("Tree", bound="TreeNode")

--- a/docs/source/whats-new.rst
+++ b/docs/source/whats-new.rst
@@ -32,6 +32,9 @@ Deprecations
 Bug fixes
 ~~~~~~~~~
 
+- Ensure nodepath class is compatible with python 3.12 (:pull:`260`)
+  By `Max Grover <https://github.com/mgrover1>`_.
+
 Documentation
 ~~~~~~~~~~~~~
 

